### ISSUE TITLE
Add documentation article about extractors. [phab:D1976]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2429]
_Created by **fromberger** at 2017-10-10 07:15:54_:

**Comment by kythe** [comment]
_**nbeloglazovgoog** wrote at 2017-10-10 07:18:43_:

I wanted to kick off extractor doc. Please suggest changes if something is wrong or I need to add something as I wrote based on my best knowledge as outside developer.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-10-10 22:17:46_:

Maybe something like: Compilation Units are captured by link:...[extractors].

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-10-10 22:17:46_:

This feels a little out of place. Maybe roll it into the previous paragraph?

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-10-10 22:17:46_:

Maybe also mention it tells how to invoke the compiler (if applicable).

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-10-10 22:17:46_:

The extractor doesn't really convert the project. I can see a description that says it takes the project as input and outputs Compilation Units. After that description, it might help to have a concrete example from bazel or maven.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-10-10 22:17:46_:

This and the following paragraph sound like the previous two paragraphs, but with more details. Can you merge the previous two paragraphs into these two?

**Comment by kythe** [comment]
_**nbeloglazovgoog** wrote at 2017-10-11 05:07:25_:

  - Address comments from review.

**Comment by kythe** [comment]
_**nbeloglazovgoog** commented inline (context lost during migration) at 2017-10-11 05:12:21_:

Refrased it as suggested.

**Comment by kythe** [comment]
_**nbeloglazovgoog** commented inline (context lost during migration) at 2017-10-11 05:12:21_:

Changed a little bit, but not much. Currently the structure is the following:

First paragraph: what is extractor

Second: why it is needed

Third: example with maven. 

Does it look reasonable?

**Comment by kythe** [comment]
_**nbeloglazovgoog** commented inline (context lost during migration) at 2017-10-11 05:12:21_:

Any other requirements for new extractors?

**Comment by kythe** [comment]
_**nbeloglazovgoog** commented inline (context lost during migration) at 2017-10-11 05:12:21_:

I don't think it belongs to the previous paragraph. The previous paragraph talks about a script that indexes the whole Kythe repo. And I want just to provide a link to all available extractors which is different from "how to index kythe repo itself".

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-10-11 13:34:05_:

For some historical context: The basic interface between the build system and the rest of Kythe is via the CompilationUnit record. That means in practice that Kythe actually comprises two separate indexing passes: One for build/dependency information, one for the code itself.

Obviously using the term "indexer" for both of these would be confusing, so we used the word "extractor" to describe the "build-system indexer" (that term mostly because of how we "extract" compilation data from Bazel).

But at root, there is nothing that requires this step to be handled to be a separate program—a build tool could be instrumented directly to capture this information (e.g., under a flag perhaps). The distinguishing feature of an "extractor" if there is one is only that it emits CompilationUnit messages and their corresponding data. What it takes as input and how it works internally are essentially unspecified by Kythe, by design.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-10-11 18:04:28_:

I'm not sure of specific requirements, especially in the way of the corpus. If it generates valid kindex files, that should be sufficient. fromberger any thoughts?

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-10-11 18:04:28_:

Why do you want the link to the list of extractors? I guess I'm just missing the motivation for putting the link here.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-10-11 18:04:28_:

It looks better. Maybe do another pass with fromberger info from above and he can give it a final OK

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-10-11 18:04:28_:

Thanks for putting this together. We can add to it over time, but it looks like a good start.

Setting review to fromberger to give a final pass.

**Comment by kythe** [comment]
_**nbeloglazovgoog** commented inline (context lost during migration) at 2017-10-12 04:24:03_:

Motivation is that users might start with this example article. It contains example with bazel extractor and users might want to see what other extractors exist because most likely they don't use bazel. That's why I thought including list to extractors here is appropriate. Otherwise, users have to search the whole kythe.io site to find the list of supported extractors.

**Comment by kythe** [comment]
_**nbeloglazovgoog** commented inline (context lost during migration) at 2017-10-12 04:24:03_:

Thanks for the context. Do you want me to add it to the doc? I understand that there are multiple ways to implement indexers. But I wanted to capture the "standard" approach which seems to be used by C++, Java, Go and JS indexers. All of them have extractors as separate programs/plugins. 

I'm happy to add another paragraph that talks about how extractor can be skipped in some cases. Do we have such indexers already? Maybe Typescript such example?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-10-12 17:39:25_:

I'm torn. On the one hand I see the value of having some more specific documentation for these tools. On the other hand I am not comfortable having a very general statement about the architecture based on what is ultimately a special case.

Writing a more general description of build-system integrations has been on my back-burner for far too long. I think I'd be more comfortable if maybe this focused only on the particulars (e.g., how to extract compilation records from Bazel or Maven) rather than talking in general about the high-level mechanism.

**Comment by kythe** [comment]
_**nbeloglazovgoog** wrote at 2017-10-13 05:33:46_:

  - Remove high-level info about indexing and keep only extractor-specific parts.

**Comment by kythe** [comment]
_**nbeloglazovgoog** commented inline (context lost during migration) at 2017-10-13 05:36:47_:

Np. Removed high-level description and left only definition of extractor and example with maven. Are you ok with having "adding new extractor" section at the bottom?

**Comment by kythe** [comment]
_**nbeloglazovgoog** wrote at 2017-10-17 12:14:30_:

Friendly ping.

**Comment by kythe** [comment]
_**fromberger** wrote at 2018-05-21 20:41:57_:

As part of the move off Phabricator to GitHub pull requests, I'm abandoning this revision. If someone wants to pursue it further, please start a new issue on https://github.com/google/kythe/issues.

I took a patch of this diff, to ensure we do not lose the content, but it will need to be updated quite a bit before we can land it.

**Comment by kythe** [comment]
_**fromberger** abandoned revision at 2018-05-21 20:41:57_
